src/context.rs

```diff
@@ -727,23 +773,24 @@ impl ContextWriter {
         let ctx = self.bc.skip_context(bo);
         self.w.symbol(skip as u32, &mut self.fc.skip_cdfs[ctx], 2);
     }
-    pub fn write_token_block_zero(&mut self, plane: usize, bo: &BlockOffset) {
+    pub fn write_token_block_zero(&mut self, plane: usize, bo: &BlockOffset, tx_size: TxSize) {
         let plane_type = if plane > 0 { 1 } else { 0 };
-        let tx_size_ctx = TXSIZE_SQR_MAP[TxSize::TX_4X4 as usize] as usize;
+        let tx_size_ctx = TXSIZE_SQR_MAP[tx_size as usize] as usize;
         let ref_type = 0;
         let band = 0;
         let ctx = self.bc.coeff_context(plane, bo);
         let cdf = &mut self.fc.coef_head_cdfs[tx_size_ctx][plane_type][ref_type][band][ctx];
         //println!("encoding token band={} ctx={}", band, ctx);
         self.w.symbol(0, cdf, HEAD_TOKENS + 1);
-        self.bc.set_coeff_context(plane, bo, false);
+        self.bc.set_coeff_context(plane, bo, tx_size, false);
     }
     pub fn write_coeffs(&mut self, plane: usize, bo: &BlockOffset,
                         coeffs_in: &[i32], tx_size: TxSize, tx_type: TxType) {
         let scan_order = &av1_intra_scan_orders[tx_size as usize][tx_type as usize];
         let scan = scan_order.scan;
-        let mut coeffs = [0 as i32; 16];
-        for i in 0..16 {
+        let mut coeffs_storage = [0 as i32; 64*64];
+        let coeffs = &mut coeffs_storage[..tx_size.width()*tx_size.height()];
+        for i in 0..tx_size.width()*tx_size.height() {
```

Comment:
``` rust
for (i, c) in coeffs.iter_mut().enumerate() {
    *c = coeffs_in[scan[i] as usize]; 
}
```

Maybe? (The variant that does coeffs.iter_mut().zip(scan.map(|s| coeffs_in[s as usize])) is less readable)

Reply:
That's definitely better. I have to update the commit anyway, will include it.